src/boot/efi/stub.c

```diff
@@ -130,6 +130,53 @@ static void export_variables(EFI_LOADED_IMAGE_PROTOCOL *loaded_image) {
         (void) efivar_set_uint64_le(LOADER_GUID, L"StubFeatures", stub_features, 0);
 }
 
+static bool use_load_options(
+                EFI_HANDLE stub_image,
+                EFI_LOADED_IMAGE_PROTOCOL *loaded_image,
+                bool have_cmdline,
+                void **ret_load_options,
+                size_t *ret_load_options_size) {
+
+        assert(stub_image);
+        assert(loaded_image);
+        assert(ret_load_options);
+        assert(ret_load_options_size);
+
+        /* We only allow custom command lines if we aren't in secure boot or if no cmdline was backed into
+         * the stub image. */
+        if (secure_boot_enabled() && have_cmdline)
+                return false;
+
+        /* We also do a superficial check whether first character of passed command line
+         * is printable character (for compat with some Dell systems which fill in garbage?). */
+        if (loaded_image->LoadOptionsSize == 0 || ((char16_t *) loaded_image->LoadOptions)[0] <= 0x1F)
+                return false;
+
+        /* The UEFI shell registers EFI_SHELL_PARAMETERS_PROTOCOL onto images it runs. This lets us know that
+         * the LoadOptions is a valid NUL-terminated string and the first arg is the stub binary path used
+         * during command execution (which we want to strip off). */
+        if (BS->HandleProtocol(
+                        stub_image,
+                        &(EFI_GUID) EFI_SHELL_PARAMETERS_PROTOCOL_GUID,
+                        &(void *){ NULL }) != EFI_SUCCESS) {
+                /* Not running from EFI shell. Use entire LoadOptions. */
+                *ret_load_options = loaded_image->LoadOptions;
+                *ret_load_options_size = loaded_image->LoadOptionsSize;
+                return true;
+        }
+
+        char16_t *rem = strchr16(loaded_image->LoadOptions, ' ');
+        if (!rem)
+                /* No arguments were provided? Then we fall back to built-in cmdline. */
+                return false;
+
+        rem++;
+        *ret_load_options = rem;
+        *ret_load_options_size = loaded_image->LoadOptionsSize -
+                        ((uint8_t *) rem - (uint8_t *) loaded_image->LoadOptions);
+        return true;
```

Comment:
hmm, wouldn't it be nicer to use the argv[] array from the shell EFI protocol instead of LoadOptions? And then join it ourselves? it sounds to me as if this was more along the line of how the protocol is supposed to be used: as the pimary input for the command line if it is specified.

hence:

1. if efi shell proto available, let's use its argv[] and ignore argv[0] and join it ourselves.
2. if efi shell proto is not available, use LoadOptions as they are